File: workers/translator/tracing.ts
import type { Context } from "@opentelemetry/api";
import { context, propagation } from "@opentelemetry/api";
import { logs } from "@opentelemetry/api-logs";
import { W3CTraceContextPropagator } from "@opentelemetry/core";
import { OTLPLogExporter } from "@opentelemetry/exporter-logs-otlp-http";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { Resource } from "@opentelemetry/resources";
import {
  BatchSpanProcessor,
  BasicTracerProvider,
  TracerConfig,
} from "@opentelemetry/sdk-trace-base";
import {
  BatchLogRecordProcessor,
  LoggerProvider,
} from "@opentelemetry/sdk-logs";

export type OpenObserveConfig = {
  endpoint: string;
  organization: string;
  authToken: string;
};

export type TraceContext = {
  traceparent?: string;
  tracestate?: string;
};

let tracerProvider: BasicTracerProvider | undefined;
let loggerProvider: LoggerProvider | undefined;

// Initializes the OTel tracer and logger providers, exporting both to
// OpenObserve via OTLP/HTTP. Must be called before any logging occurs.
export function initTelemetry(cfg: OpenObserveConfig) {
  const resource = new Resource({
    ["service.name"]: "epub-web-tool-translator",
  });

  const spanProcessor = new BatchSpanProcessor(
    new OTLPTraceExporter({
      url: `http://${cfg.endpoint}/api/${cfg.organization}/v1/traces`,
      headers: {
        Authorization: `Basic ${cfg.authToken}`,
      },
    })
  );

  tracerProvider = new BasicTracerProvider({
    resource,
    spanProcessors: [spanProcessor],
  });

  tracerProvider.register();
  propagation.setGlobalPropagator(new W3CTraceContextPropagator());

  loggerProvider = new LoggerProvider({ resource });
  loggerProvider.addLogRecordProcessor(
    new BatchLogRecordProcessor(
      new OTLPLogExporter({
        url: `http://${cfg.endpoint}/api/${cfg.organization}/v1/logs`,
        headers: { Authorization: `Basic ${cfg.authToken}` },
      })
    )
  );
  logs.setGlobalLoggerProvider(loggerProvider);

  return {
    shutdown: async () => {
      await tracerProvider?.shutdown();
      await loggerProvider?.shutdown();
    },
  };
}

// TraceParentCarrier lets the W3C propagator read/write the trace context that
// travels inside the queue message JSON body.
export class TraceParentCarrier {
  constructor(private readonly traceContext: TraceContext) {}

  get(key: string): string | undefined {
    switch (key) {
      case "traceparent":
        return this.traceContext.traceparent;
      case "tracestate":
        return this.traceContext.tracestate;
      default:
        return undefined;
    }
  }

  set(key: string, value: string): void {
    switch (key) {
      case "traceparent":
        this.traceContext.traceparent = value;
        break;
      case "tracestate":
        this.traceContext.tracestate = value;
        break;
    }
  }

  keys(): string[] {
    return Object.keys(this.traceContext);
  }
}

// extractParentContext rebuilds a remote parent context from the serialized
// W3C trace context embedded in a queue message.
export function extractParentContext(traceContext: TraceContext): Context {
  return propagation.extract(
    context.active(),
    new TraceParentCarrier(traceContext)
  );
}

// encodeTraceContext writes the trace context of ctx into the carrier-backed
// object and returns it, so it can travel in the queue message body.
export function encodeTraceContext(
  ctx: Context,
  traceContext: TraceContext
): TraceContext {
  propagation.inject(ctx, new TraceParentCarrier(traceContext));
  return traceContext;
}

File: internal/otel/tracing.go
package otel

import (
	"context"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/trace"
)

func Tracer(name string) trace.Tracer {
	return otel.Tracer(name)
}

// EncodeTraceParent serializes the W3C trace context (traceparent/tracestate)
// from the active span in ctx. Returns empty strings when there is no valid span.
func EncodeTraceParent(ctx context.Context) (traceparent string, tracestate string) {
	span := trace.SpanFromContext(ctx)
	if !span.SpanContext().IsValid() {
		return "", ""
	}
	carrier := &traceCarrier{}
	otel.GetTextMapPropagator().Inject(ctx, carrier)
	return carrier.Get("traceparent"), carrier.Get("tracestate")
}

// ContextFromTraceParent rebuilds a parent context from a serialized W3C
// traceparent/tracestate. The returned context carries no active span; callers
// use it as the parent of a new consumer span.
func ContextFromTraceParent(ctx context.Context, traceparent string, tracestate string) context.Context {
	if traceparent == "" {
		return ctx
	}
	return otel.GetTextMapPropagator().Extract(ctx, &traceCarrier{
		traceparent: traceparent,
		tracestate:  tracestate,
	})
}

type traceCarrier struct {
	traceparent string
	tracestate  string
}

func (c *traceCarrier) Get(key string) string {
	switch key {
	case "traceparent":
		return c.traceparent
	case "tracestate":
		return c.tracestate
	}
	return ""
}

func (c *traceCarrier) Set(key, value string) {
	switch key {
	case "traceparent":
		c.traceparent = value
	case "tracestate":
		c.tracestate = value
	}
}

func (c *traceCarrier) Keys() []string {
	keys := make([]string, 0, 2)
	if c.traceparent != "" {
		keys = append(keys, "traceparent")
	}
	if c.tracestate != "" {
		keys = append(keys, "tracestate")
	}
	return keys
}

func SetSpanAttributes(ctx context.Context, attrs ...attribute.KeyValue) {
	if span := trace.SpanFromContext(ctx); span.SpanContext().IsValid() {
		span.SetAttributes(attrs...)
	}
}

func RecordError(ctx context.Context, err error) {
	if err != nil {
		trace.SpanFromContext(ctx).RecordError(err)
	}
}

var (
	_ propagation.TextMapCarrier = (*traceCarrier)(nil)
)
